libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/OnlineRouteAlternativesSwitch.kt

```diff
@@ -0,0 +1,254 @@
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.mapLatest
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.withContext
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import java.util.concurrent.TimeUnit
+import kotlin.coroutines.resume
+
+private const val LOG_CATEGORY = "OnlineRouteAlternativesSwitch"
+
+/***
+ * This components replaces offline route by an online alternative when it's available. Under the
+ * hood it's requesting routes in cycle from the current position until the first successful result.
+ *
+ * Normally this job should be done by `RouteAlternativesObserver`. However some environments
+ * doesn't have working reachability/connectivity API. This class is a temporary workaround for
+ * those environments until we reimplement route alternatives logic handle offline-online switch
+ * even if platform's reachability API doesn't work.
+ *
+ * @param connectTimeoutMilliseconds - is forwarded to OkHttp as is
+ * @param readTimeoutMilliseconds - is forwarded to OkHttp as is
+ * @param minimumRetryInterval - defines an interval that limits how often routes will be requested.
+ * Route requests won't happen more often than this interval.
+ */
+@ExperimentalMapboxNavigationAPI
+class OnlineRouteAlternativesSwitch(
+    private val connectTimeoutMilliseconds: Int = 10_000,
+    private val readTimeoutMilliseconds: Int = 30_000,
+    private val minimumRetryInterval: Int = 60_000,
+) : MapboxNavigationObserver {
+
+    private lateinit var mapboxNavigationScope: CoroutineScope
+
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        mapboxNavigationScope = CoroutineScope(Dispatchers.Main + SupervisorJob())
+        val accessToken = mapboxNavigation.navigationOptions.accessToken ?: return
+        mapboxNavigationScope.launch {
+            mapboxNavigation.flowRoutesUpdated()
+                .map { it.navigationRoutes }
+                .requestOnlineRoutes(
+                    mapboxNavigation.flowLocationMatcherResult(),
+                    mapboxNavigation.flowRouteProgress(),
+                    routeRequestMechanism = { options ->
+                        requestRoutes(
+                            options,
+                            accessToken,
+                            MapboxNavigationAccounts,
+                            connectTimeoutMilliseconds = connectTimeoutMilliseconds,
+                            readTimeoutMilliseconds = readTimeoutMilliseconds
+                        )
+                    },
+                    minimumRetryInterval = minimumRetryInterval.toLong()
+                )
+                .collectLatest {
+                    mapboxNavigation.setNavigationRoutes(it)
```

Comment:
Was it decided we should proactively set new routes and completely discard the old ones in this case? We don't do that for alternatives.

Reply:
I'm waiting for feedback from target customer

Reply:
The target customer is happy about automatic set routes